### ISSUE TITLE
Private AKS cluster fails with 'error getting service account token'

### ISSUE THREAD

**Issue by annablender** [rancher/rancher#37164]
**Rancher Server Setup**
- Rancher version: 2.6.3
- Installation option (Docker install/Helm Chart): Hosted Rancher
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc):
- Proxy/Cert Details:

**Information about the Cluster**
- Kubernetes version: 
- Cluster Type (Local/Downstream): Private AKS cluster, 3 worker nodes
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider):

   
**Describe the bug**

After creating a private AKS cluster in Rancher (selecting the 'Enable Private Cluster'), the cluster is provisioned. Following through and running the import command (via /g endpoint) to apply the agent manifest, the logs of the cluster-agent look normal, and Rancher logs a ‘handling backend connection’ when the agent started. However, Rancher appears to be persistently attempting a post to the privatelink endpoint, preventing the cluster from becoming active.

```
Example:
2022/03/21 22:05:01 [ERROR] error syncing 'c-dm284': handler aks-operator-controller: error getting service account token: Post "https://aks-dev-canadacentral-fadc97fa.9dd212f6-4e68-493e-b950-44a153eb3845.privatelink.canadacentral.azmk8s.io:443/api/v1/namespaces": dial tcp: address aks-dev-canadacentral-fadc97fa.9dd212f6-4e68-493e-b950-44a153eb3845.privatelink.ca, requeuing
```

**To Reproduce**
 

David reproduced this issue in a hosted rancher QA lab. Cluster has connectivity with the agent via web socket but the cluster is not active.

His logs/error

Apply agent manifest:
```
az aks command invoke --resource-group dnoland-aks --name dnoland-aks-privatelink  --command "kubectl apply -f https://support.rancher.cloud/v3/import/prgnb6t4qbf54xpc29zfp46bkdfhrjx4jvs5lt4vg8954q68clvsqv_c-jr6cc.yaml"
command started at 2022-03-22 16:55:47+00:00, finished at 2022-03-22 16:55:48+00:00 with exitcode=0
clusterrole.rbac.authorization.k8s.io/proxy-clusterrole-kubeapiserver created
clusterrolebinding.rbac.authorization.k8s.io/proxy-role-binding-kubernetes-master created
namespace/cattle-system created
serviceaccount/cattle created
clusterrolebinding.rbac.authorization.k8s.io/cattle-admin-binding created
secret/cattle-credentials-6303327 created
clusterrole.rbac.authorization.k8s.io/cattle-admin created
deployment.apps/cattle-cluster-agent created
service/cattle-cluster-agent created
```

cluster-agent:
```
INFO: Environment: CATTLE_ADDRESS=10.244.0.9 CATTLE_CA_CHECKSUM= CATTLE_CLUSTER=true CATTLE_CLUSTER_AGENT_PORT=tcp://10.0.235.71:80 CATTLE_CLUSTER_AGENT_PORT_443_TCP=tcp://10.0.235.71:443 CATTLE_CLUSTER_AGENT_PORT_443_TCP_ADDR=10.0.235.71 CATTLE_CLUSTER_AGENT_PORT_443_TCP_PORT=443 CATTLE_CLUSTER_AGENT_PORT_443_TCP_PROTO=tcp CATTLE_CLUSTER_AGENT_PORT_80_TCP=tcp://10.0.235.71:80 CATTLE_CLUSTER_AGENT_PORT_80_TCP_ADDR=10.0.235.71 CATTLE_CLUSTER_AGENT_PORT_80_TCP_PORT=80 CATTLE_CLUSTER_AGENT_PORT_80_TCP_PROTO=tcp CATTLE_CLUSTER_AGENT_SERVICE_HOST=10.0.235.71 CATTLE_CLUSTER_AGENT_SERVICE_PORT=80 CATTLE_CLUSTER_AGENT_SERVICE_PORT_HTTP=80 CATTLE_CLUSTER_AGENT_SERVICE_PORT_HTTPS_INTERNAL=443 CATTLE_CLUSTER_REGISTRY= CATTLE_INGRESS_IP_DOMAIN=sslip.io CATTLE_INSTALL_UUID=c5ebd2f9-fcd5-4c85-a88a-44471fe98afa CATTLE_INTERNAL_ADDRESS= CATTLE_IS_RKE=false CATTLE_K8S_MANAGED=true CATTLE_NODE_NAME=cattle-cluster-agent-64d5f97fbb-2ghcf CATTLE_SERVER=https://support.rancher.cloud CATTLE_SERVER_VERSION=v2.6.3
INFO: Using resolv.conf: search cattle-system.svc.cluster.local svc.cluster.local cluster.local ms1rtotrucqutpzifvx3lszrre.bx.internal.cloudapp.net nameserver 10.0.0.10 options ndots:5
INFO: https://support.rancher.cloud/ping is accessible
INFO: support.rancher.cloud resolves to 54.201.237.77 54.69.62.197
time="2022-03-22T16:56:09Z" level=info msg="Listening on /tmp/log.sock"
time="2022-03-22T16:56:09Z" level=info msg="Rancher agent version v2.6.3 is starting"
time="2022-03-22T16:56:09Z" level=info msg="Connecting to wss://support.rancher.cloud/v3/connect/register with token starting with prgnb6t4qbf54xpc29zfp46bkdf"
time="2022-03-22T16:56:09Z" level=info msg="Connecting to proxy" url="wss://support.rancher.cloud/v3/connect/register"
```
 
Error from cluster status:
```
failed to communicate with cluster: error getting service account token: Post "https://dnoland-0eacef28.57bbc7b8-6ec4-4b9f-b298-d99425eadff2.privatelink.eastus.azmk8s.io:443/api/v1/namespaces": cluster agent disconnected  
```

**Result**
Cluster is not active

**Expected Result**

Cluster should come up as active

SURE-4280

**Comment by annablender** [comment]
I was able to reproduce this issue on 2.6-head this morning.
 
Local Rancher server: dev rainbow cluster (non-hosted Rancher)
Cluster type: Custom AKS downstream
Nodes: 3
PrivateCluster? Yes
Notes: Right after the AKS cluster finished provisioning, the service token error showed up in the Rancher UI.

`RequiresPrivateTunnel: true` is set.

There's no errors in the cattle-agent logs. Because the Explorer button is true, that means the cattle-agent has successfully connected with the Rancher server. Because the `privateRequiresTunnel` is set to true, that means Rancher has determined it is outside the private cluster's vent and requires a tunnel to the connect to the downstream cluster. This indicates a backend problem with the tunnel, possibly in the tunnel dialer. Further investigation is required. See the above partial workaround to fix the connection if Rancher cannot detect the vnet.

**Comment by annablender** [comment]
The way that `ask-operator` deals with auto-generated tagging in AKS is a bug. I've opened a new issue for it here https://github.com/rancher/rancher/issues/37608.

**Comment by annablender** [comment]
## Testing Template

Please test this issue by following [these repro steps](https://github.com/rancher/rancher/issues/37608#issuecomment-1137924654) but with a private AKS cluster on `2.6.3` and `2.6-head.`

**Comment by markusewalker** [comment]
Verified on `v2.6-head` https://github.com/rancher/rancher/commit/e93864b3364e18e81dfd5c12366206f2d495cf52.

**ENVIRONMENT DETAILS**
- Rancher HA
- Rancher version
     - Reproduced version: `v2.6.5`
     - Verified version: `v2.6-head` https://github.com/rancher/rancher/commit/e93864b3364e18e81dfd5c12366206f2d495cf52
	• Browser: Chrome

**TEST RESULT**
PASS

**VERIFICATION STEPS**
1. In a local client machine, ensure that you have downloaded the `az` cli and have logged into your Azure subscription.
2. Setup Rancher and navigate to the Rancher UI.
3. Create a standard user and login to Rancher as that user.
4. Create a 3 node, private AKS cluster via Rancher. Be sure to set the following:
     - Check off `Enable Private Cluster` in `Networking Options`.
     - In `Cluster Options`, add tags that match your Kubernetes service. If you know them beforehand, you can do it before you create it. If not, you will need to come back to this step and add the tags.
5. Navigate back to your client machine and run the following command once the cluster is created: `az aks command invoke --resource-group <resource group> --name <cluster name> --command "<registration command"
6. Verified the cluster is up and active:
![image](https://user-images.githubusercontent.com/52591819/170602854-e4fa9580-f992-4734-b293-562a89b9a5dc.png) 

A concern that I see is that there is a constant status change as illustrated in the below gif.
![test](https://user-images.githubusercontent.com/52591819/170603014-0bf5bed8-734b-42cf-be19-d0d34afa72c1.gif)

I believe this is more so environmental to my networking configuration with my cluster. In the backend, the `aks-command` namespace is constantly going into a status of `Completed`, spins up a new namespace until it goes into `Running`, and then repeats this over and over. The visual result is the gif above. Unsure if this is an unexpected regression or more environmental to my cluster as mentioned.

@annablender before this issue is closed, let's connect on this to verify my suspicions.

**Comment by markusewalker** [comment]
Reopening this issue. Despite the original issue of the `error getting service account token` being gone, the cluster status is not in a solid `Active` state. As such, we should close this issue as it is customer reported and closing gives off the impression it is fully addressed.

As https://github.com/rancher/rancher/issues/37855 is essentially a duplicate of this current issue, I closed that one to keep our eyes on this ticket.

**Comment by annablender** [comment]
The aks operator version v1.0.6-rc2 has the [tags sync fix](https://github.com/rancher/aks-operator/releases/tag/v1.0.6-rc2) and rancher is already running past that version so `rancher 2.6-head` has the tags fix, but it appears to have not fixed this particular problem. Doing further investigation now

**Comment by annablender** [comment]
When re-investigating this issue, I ran into a bug in the rancher aks handler where if the service account secret value returns as an empty string from the `SecretsMigrator` code, rancher will panic when trying to reference `secret.Name.` This has been fixed here https://github.com/rancher/rancher/blob/9b3c5d72d02659fd7280260a178162ebdd6a4507/pkg/controllers/management/aks/aks_cluster_handler.go#L201 and target release is 2.6.7. The secret is returning empty because the cattle namespace fails to create. Still looking into the root cause

**Comment by annablender** [comment]
After merging the panic fix, I am not seeing what @markusewalker saw [here](https://github.com/rancher/rancher/issues/37164#issuecomment-1139161978) with the flickering icon on a private 1 node AKS cluster. The cattle system namespace does not recreate.

Steps
Repro steps
* Run rancher 2.6-head locally
* Login
* Add Azure credentials
* Create 3 node private AKS cluster with kubernetes 1.23.5
* Setup a [bastion vm](https://docs.microsoft.com/en-us/azure/virtual-machines/linux/quick-create-portal) in the same vnet as downstream cluster.
* Add AllowInBoundPort22 rule to networking rules
* Grab the private cluster's kubeconfig

`az aks get-credentials --resource-group <resource group> --name <cluster name> --file <kubeconfig file name>`

* SSH into the bastion

`ssh -I <path to secret key> azureuser@<virtual machine IP>`

* Install kubectl, copy the kubeconfig into ~/.kube/config, and run the import command

![image](https://user-images.githubusercontent.com/12955547/178360141-490997ce-aba4-4ea4-835d-397c38e1351f.png)

![image](https://user-images.githubusercontent.com/12955547/178360160-16889e3e-33bf-4e9a-b847-614b3b87f320.png)

**Comment by markusewalker** [comment]
Verified that this is addressed in `v2.6-head` https://github.com/rancher/rancher/commit/3506e43cee0d391bd3fdc32d3d98690b8abef402.

**ENVIRONMENT DETAILS**
- Rancher Install: Helm
- Rancher version: `v2.6-head` https://github.com/rancher/rancher/commit/3506e43cee0d391bd3fdc32d3d98690b8abef402
- Browser: Chrome

**TEST RESULT**
PASS

**VERIFICATION STEPS**
1. Setup Rancher and navigated to the Rancher UI.
2. Create a 3 node, private AKS cluster via Rancher. Be sure to set the following:
     - Check off `Enable Private Cluster` in `Networking Options`.
3. Created an Azure VM and ensured that `az` and `kubectl` are installed on the machine.
     - Made sure that the VM is in the same vnet as the private cluster, so I could directly communicate to the cluster.
4. Once the cluster was provisioned in the Azure Portal, navigated to the VM and ran the registration command.
5. Verified the cluster is up and active and DOES NOT flicker statuses as seen before:
![image](https://user-images.githubusercontent.com/52591819/178545934-0bf6fc02-e408-460b-994a-8c878f216b02.png)
![image](https://user-images.githubusercontent.com/52591819/178545949-8a269347-0642-463a-b61d-45a016ca540c.png)
![image](https://user-images.githubusercontent.com/52591819/178545963-488ecab5-6436-4886-b5c2-665b3efa3180.png)
6. Navigated to the VM and ran `kubectl get pods -A`; verified that no pods are being constantly terminated as seen before. I see the expected pods:

```
$ kubectl get pods -A
NAMESPACE       NAME                                    READY   STATUS    RESTARTS      AGE
cattle-system   cattle-cluster-agent-555f79b6f5-cc7lt   1/1     Running   0             15m
cattle-system   cattle-cluster-agent-555f79b6f5-t5wdr   1/1     Running   0             15m
kube-system     azure-ip-masq-agent-2tr2x               1/1     Running   0             18m
kube-system     azure-ip-masq-agent-lfr4p               1/1     Running   0             18m
kube-system     azure-ip-masq-agent-tb9rj               1/1     Running   0             18m
kube-system     cloud-node-manager-5xnvs                1/1     Running   0             18m
kube-system     cloud-node-manager-9g2mz                1/1     Running   0             18m
kube-system     cloud-node-manager-rnqk7                1/1     Running   0             18m
kube-system     coredns-87688dc49-74sw9                 1/1     Running   0             22m
kube-system     coredns-87688dc49-9wpnk                 1/1     Running   0             22m
kube-system     coredns-autoscaler-6fb889cdfc-s9fmd     1/1     Running   0             22m
kube-system     csi-azuredisk-node-4btd8                3/3     Running   0             18m
kube-system     csi-azuredisk-node-8fcqd                3/3     Running   0             18m
kube-system     csi-azuredisk-node-ztnpv                3/3     Running   0             18m
kube-system     csi-azurefile-node-6w9mq                3/3     Running   0             18m
kube-system     csi-azurefile-node-ktb5m                3/3     Running   0             18m
kube-system     csi-azurefile-node-vzxt5                3/3     Running   0             18m
kube-system     konnectivity-agent-6bd7b8bfb6-wkqt4     1/1     Running   0             22m
kube-system     konnectivity-agent-6bd7b8bfb6-xz4p5     1/1     Running   0             22m
kube-system     kube-proxy-58zm7                        1/1     Running   0             18m
kube-system     kube-proxy-vrcsd                        1/1     Running   0             18m
kube-system     kube-proxy-z75m2                        1/1     Running   0             18m
kube-system     metrics-server-948cff58d-f2fxq          1/1     Running   1 (17m ago)   22m
kube-system     metrics-server-948cff58d-t5zl2          1/1     Running   1 (17m ago)   22m
```
